sympy/sets/handlers/intersection.py

```diff
@@ -235,26 +235,44 @@ def intersection_sets(self, other): # noqa:F811
     # diophantine equations f(n)=g(m).
     # If the solutions for n are {h(t) : t in Integers} then we return
     # {f(h(t)) : t in integers}.
+    # If the solutions for n are {n_1, n_2, ..., n_k} then we return
+    # {f(n_i) : 1 <= i <= k}.
     if base_set is S.Integers:
         gm = None
         if isinstance(other, ImageSet) and other.base_sets == (S.Integers,):
             gm = other.lamda.expr
-            m = other.lamda.variables[0]
+            var = other.lamda.variables[0]
+            # Symbol of second ImageSet lambda must be distinct from first
+            m = Dummy('m')
+            gm = gm.subs(var, m)
         elif other is S.Integers:
-            m = gm = Dummy('x')
+            m = gm = Dummy('m')
         if gm is not None:
             fn = self.lamda.expr
             n = self.lamda.variables[0]
-            solns = list(diophantine(fn - gm, syms=(n, m)))
+            try:
+                solns = list(diophantine(fn - gm, syms=(n, m), permute=True))
+            except (TypeError, NotImplementedError):
+                return
```

Comment:
What raises `TypeError`?

Reply:
https://github.com/sympy/sympy/blob/c559a8421ac4865ebfe66024be6cd43a6103a62b/sympy/solvers/diophantine.py#L202-L204
That's the exception raised by `diophantine` that was at the core of a number of filed issues. `diophantine` does a good job of checking if it can handle the passed equation and this is how it complains about one specific type of "misuse". (Whereas `NotImplementedError` is raised if the equation would be fine, but it simply doesn't know how to solve it.)

Reply:
I think that should be `ValueError` rather than `TypeError`. I also very much dislike the way that the diophantine function uses large try blocks. Catching `AttributeError` shouldn't happen anywhere...

Reply:
I just tried this PR with this diff:
```diff
diff --git a/sympy/sets/handlers/intersection.py b/sympy/sets/handlers/intersection.py
index 848cf2ee10..55c7ec35bd 100644
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -252,7 +252,7 @@ def intersection_sets(self, other): # noqa:F811
             n = self.lamda.variables[0]
             try:
                 solns = list(diophantine(fn - gm, syms=(n, m), permute=True))
-            except (TypeError, NotImplementedError):
+            except NotImplementedError:
                 return
             # 3 cases are possible for solns:
             # - empty set,
```
I ran all non-slow tests in sets and solvers and didn't get any failures...

Reply:
> I also very much dislike the way that the `diophantine` function uses large try blocks

Agreed. I was astonished to see such a large `try` block when I investigated #18196 a bit. (And that `try` block even contributed to it.)

As for the `TypeError`, it should be sufficient to try the one from #18018 
```
imageset(Lambda(n, n*log(2)), S.Integers).intersection(S.Integers)
```

Reply:
I'll add a test, simply to exclude that this execption is removed from the list.

Reply:
Can you also add a comment explaining where the exceptions are raised?

The biggest problem with cleaning up all the exception around the codebase is not knowing what it potentially catches or what the author intended it to catch.

Reply:
Sure, I'll add a note.

Reply:
As far as I can tell from the tests the `AttributeError` is just there to handle this case:
```julia
In [1]: diophantine(3)                                                                                                            
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/current/sympy/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    174     try:
--> 175         var = list(eq.expand(force=True).free_symbols)
    176         var.sort(key=default_sort_key)

AttributeError: 'int' object has no attribute 'expand'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-1-50971d7489ec> in <module>
----> 1 diophantine(3)

~/current/sympy/sympy/sympy/solvers/diophantine.py in diophantine(eq, param, syms, permute)
    201         assert eq.is_polynomial()
    202     except (GeneratorsNeeded, AssertionError, AttributeError):
--> 203         raise TypeError(filldedent('''
    204     Equation should be a polynomial with Rational coefficients.'''))
    205 

TypeError: 
Equation should be a polynomial with Rational coefficients.
```
However the sympified version works fine:
```julia
In [3]: diophantine(S(3))                                                                                                         
Out[3]: set()
```
The error is just that diophantine doesn't `eq = _sympify(eq)` so the only effect of catching `AttributeError` is to cover up that bug.